lib/spack/spack/concretize.py

```diff
@@ -203,46 +202,101 @@ def concretize_version(self, spec):
+            default = "default_os"
+        else:
+            other_comp = "platform_os"
+            default = "default_target"
 
-        if spec.architecture is None:
-            spec.architecture = spack.spec.ArchSpec(sys_arch)
-            spec_changed = True
+        def is_front_end(arch_comp):
+            fe_arch = str(getattr(platform, methods[arch_comp])(front_end))
+            if getattr(architecture, arch_comp) == fe_arch:
+                    return True
+            return False
+
+        if architecture.concrete or this_component:
+            return False
 
-        default_archs = list(x for x in [root_arch, sys_arch] if x)
-        for arch in default_archs:
-            if spec.architecture.concrete:
-                break
+        if getattr(architecture, other_comp):
+            if is_front_end(other_comp):
+                setattr(architecture, component,
+                        getattr(platform, methods[component])(front_end))
+            else:
+                setattr(architecture, component,
+                        getattr(platform, methods[component])(default))
+        return True  # spec changed
 
-            replacement_fields = [k for k, v in iteritems(arch.to_cmp_dict())
-                                  if v and not getattr(spec.architecture, k)]
-            for field in replacement_fields:
-                setattr(spec.architecture, field, getattr(arch, field))
-                spec_changed = True
+    def concretize_architecture(self, spec):
+        """If the spec's architecture is already concrete then return False.
+        If the spec is missing the architecture portion of the spec then look
+        through the DAG and find a spec with an architecture. If such a spec
+        exists then assign it to the current spec, otherwise provide the
+        defaults for the system. If the architecture is a Arch object then
+        check if either the platform, target, or operating system are
+        concretized. If any of the fields are changed return True."""
+        sys_arch = spack.spec.ArchSpec(spack.architecture.sys_type())
+        front_arch = spack.spec.ArchSpec(spack.architecture.front_end())
 
-        if not spec.architecture.concrete:
-            raise InsufficientArchitectureInfoError(spec, default_archs)
+        if spec.architecture and spec.architecture.concrete:
+            return False
 
-        return spec_changed
+        spec_changed = []
+        other_arch_spec = find_spec(spec, lambda x: x.architecture, spec.root)
+        other_arch = other_arch_spec.architecture
+        if not other_arch:
+            spec.architecture = spack.spec.ArchSpec(sys_arch)
+        elif not spec.architecture and other_arch:
+            spec.architecture = other_arch
+            spec_changed.append(True)
+
+        platform = spack.architecture.get_platform(spec.architecture.platform)
+
+        spec_changed.append(self._concretize_arch_component("platform_os",
```

Comment:
Nitpick: I'd prefer this was laid out like

```
spec_changed.append(secl._concretize_arch_component(
  "platform_os", platform, spec))
```

(although above I also mention I think `_concretize_arch_component` could be simplified to omit the `component` parameter)